### ISSUE TITLE
False positive `used-before-assignment` with dataclasses and shadowing

### ISSUE THREAD

**Issue by The-Compiler** [PyCQA/pylint#6748]
### Bug description

This file:

```python
from dataclasses import dataclass
import collections


@dataclass
class Shadow:

    collections: collections.deque
    other: collections.defaultdict


print(Shadow(collections.deque(), collections.defaultdict()))
```

is perhaps a bit unorthodox, but runs fine, because the `collections: ...` doesn't actually shadow the module when defining `other`.

### Command used

```shell
pylint -E test.py
```

### Pylint output

```shell
************* Module test
.../test.py:9:11: E0601: Using variable 'collections' before assignment (used-before-assignment)
```

### Expected behavior

No error.

### Pylint version

```shell
pylint 2.13.9
astroid 2.11.5
Python 3.10.4 (main, Mar 23 2022, 23:05:40) [GCC 11.2.0]
```

### OS / Environment

Archlinux
